### PR TITLE
split off expressionutil.d from expression.d

### PR DESCRIPTION
**Pull request by WalterBright**
This is a crude cut&paste job to move the non-AST code from expression.d to expressionutil.d. I'll save the finesse for a later PR. The code hasn't changed at all, it's just moved.

Moving this direction was a request from Luis, and it was already in my queue of things to do.

### CODE REVIEW

File: compiler/src/dmd/expressionutil.d [hunk 1]
```diff
@@ -0,0 +1,1057 @@
...
+ * Regard NaN's as equivalent.
+ * Regard +0 and -0 as different.
+ * Params:
+ *      x1 = first operand
+ *      x2 = second operand
+ * Returns:
+ *      true if x1 is x2
+ *      else false
+ */
+bool RealIdentical(real_t x1, real_t x2) @safe
```
Comment (thewilsonator) [inline compiler/src/dmd/expressionutil.d:562]: I feel this should be a final (private static) member of `RealExp` as it is only used by RealExp and ComplexExp

File: compiler/src/dmd/expressionutil.d [hunk 2]
```diff
@@ -0,0 +1,1057 @@
...
+        }
+    }
+    return true;
+}
+
+/******************************
+ * check e is exp.opDispatch!(tiargs) or not
+ * It's used to switch to UFCS the semantic analysis path
+ */
+bool isDotOpDispatch(Expression e)
```
Comment (thewilsonator) [inline compiler/src/dmd/expressionutil.d:231]: this function appears completely unused.

File: compiler/src/dmd/expressionutil.d [hunk 3]
```diff
@@ -0,0 +1,1057 @@
...
+            (*exps)[i] = Expression.combine(te.e0, (*exps)[i]);
+            arg = (*exps)[i];
+        }
+    }
+}
+
+/****************************************
+ * Expand alias this tuples.
+ */
+TupleDeclaration isAliasThisTuple(Expression e)
```
Comment (thewilsonator) [inline compiler/src/dmd/expressionutil.d:338]: this has a single use in `dsymbolsem.d`, it should be a private function there

File: compiler/src/dmd/expressionutil.d [hunk 4]
```diff
@@ -0,0 +1,1057 @@
...
+        }
+    }
+    return null;
+}
+
+/************************************************
+ * If we want the value of this expression, but do not want to call
+ * the destructor on it.
+ */
+Expression valueNoDtor(Expression e)
```
Comment (thewilsonator) [inline compiler/src/dmd/expressionutil.d:435]: this is only used in `statementsem.d` it should be private there

File: compiler/src/dmd/expressionutil.d [hunk 5]
```diff
@@ -0,0 +1,1057 @@
...
+/***************************************************
+ * Given an Expression, find the variable it really is.
+ *
+ * For example, `a[index]` is really `a`, and `s.f` is really `s`.
+ * Params:
+ *      e = Expression to look at
+ * Returns:
+ *      variable if there is one, null if not
+ */
+VarDeclaration expToVariable(Expression e)
```
Comment (thewilsonator) [inline compiler/src/dmd/expressionutil.d:588]: this s only used in `escape.d` it should be a private function there

**Review by thewilsonator — CHANGES_REQUESTED** [review]
I don't love this (we need less files, or better organised files), but what concerns me is this file isn't either clearly AST nor is it semantic. It's _mostly_ semantic with a few misc function and other definitions:

As a start, could you please (preferably as a separate PR) move those functions that should obviously be private to their respective modules and delete the one used function.

File: compiler/src/dmd/dmodule.d [hunk 6]
```diff
@@ -1626,7 +1626,7 @@ private const(char)[] processSource (const(ubyte)[] src, Module mod)
     // It's UTF-8
     if (buf[0] >= 0x80)
     {
-	auto loc = mod.getLoc();
```
[the remark this answers is not in the document]
Comment (WalterBright) [reply compiler/src/dmd/dmodule.d:1629]: `error` wants a ref argument, not a value.

File: compiler/src/dmd/dtemplate.d [hunk 7]
```diff
@@ -5629,7 +5630,7 @@ extern (C++) final class TemplateValueParameter : TemplateParameter
                     // i.e: `template T(int arg = T)`
                     // Raise error now before calling resolveProperties otherwise we'll
                     // start looping on the expansion of the template instance.
-		    auto td = sc.tinst.tempdecl;
+                    auto td = sc.tinst.tempdecl;
```
[the remark this answers is not in the document]
Comment (WalterBright) [reply compiler/src/dmd/dtemplate.d:5633]: Before I check in, I convert tabs to spaces. Sometimes it catches stray tabs in other files.

**Comment by WalterBright** [comment]
I was unclear about the purpose of this. It's not to move the free functions into other files. It's about moving the **AST** into another file. The confusing part is I wanted to name the file the AST was moved to "expression.d" to keep consistency with "statement.d", etc.

Putting it in its own file is a move towards combining the Expression AST with the one from astbase.d, which will make dmd-as-library a much more tractable problem.

While your suggestions are good, they are more suitable for follow-on. Since this PR is rather tectonic, to make it reviewable I restricted it to being a cut&paste job as much as possible.

File: compiler/src/dmd/expressionutil.d (hunk 2 above)
Comment (thewilsonator) [reply compiler/src/dmd/expressionutil.d:231]: #15685

File: compiler/src/dmd/expressionutil.d (hunk 3 above)
Comment (thewilsonator) [reply compiler/src/dmd/expressionutil.d:338]: #15686

File: compiler/src/dmd/expressionutil.d (hunk 4 above)
Comment (thewilsonator) [reply compiler/src/dmd/expressionutil.d:435]: #15687

**Comment by thewilsonator** [comment]
> While your suggestions are good, they are more suitable for follow-on.

On the contrary, given the large volume of stuff touched by this PR that should go elsewhere it is not (yet) obvious that a new file should be created. Until such a time as DMD is packaged, adding files willy-nilly is not something I approve of. Once all the other PRs I have created for this have been merged and if there is still suff left over that it is unclear where it should go, then we should consider adding a new file.

File: compiler/src/dmd/expressionutil.d (hunk 5 above)
Comment (thewilsonator) [reply compiler/src/dmd/expressionutil.d:588]: #15688

File: compiler/src/dmd/expressionutil.d [hunk 8]
```diff
@@ -0,0 +1,1057 @@
...
+
+/***********************************
+ * Determine if a `this` is needed to access `d`.
+ * Params:
+ *      sc = context
+ *      d = declaration to check
+ * Returns:
+ *      true means a `this` is needed
+ */
+bool isNeedThisScope(Scope* sc, Declaration d)
```
[the remark this answers is not in the document]
Comment (thewilsonator) [reply compiler/src/dmd/expressionutil.d:196]: #15689

File: compiler/src/dmd/expressionutil.d (hunk 1 above)
Comment (thewilsonator) [reply compiler/src/dmd/expressionutil.d:562]: #15690

**Comment by WalterBright** [comment]
Putting the Expression AST in its own module is central to merging it with astbase's implementation, and moving forward to D as a library. I don't see another way forward.

File: compiler/src/dmd/expressionutil.d [hunk 9]
```diff
@@ -0,0 +1,1057 @@
...
+/***************************************
+ * Parameters:
+ *      sc:     scope
+ *      flag:   1: do not issue error message for invalid modification
+                2: the exp is a DotVarExp and a subfield of the leftmost
+                   variable is modified
+ * Returns:
+ *      Whether the type is modifiable
+ */
+extern(D) Modifiable checkModifiable(Expression exp, Scope* sc, ModifyFlags flag = ModifyFlags.none)
```
[the remark this answers is not in the document]
Comment (RazvanN7) [reply compiler/src/dmd/expressionutil.d:721]: With this one I agree that its place is in `expressionsem.d`

**Comment by RazvanN7** [comment]
> Putting the Expression AST in its own module is central to merging it with astbase's implementation, and moving forward to D as a library. I don't see another way forward.

Yes, to that end we have `expression.d` which should be trimmed to have only the Expression AST and some helper methods that are usable at parse time and everything that requires semantic information (e.g. functions that requires `import expressionsem;` or `import *sem;`) should be moved to their corresponding semantic file. I don't see the point in creating another file called `expressionutil.d`

**Comment by RazvanN7** [comment]
To be clear, at the end of the process, `expression.d` should only contain imports to other AST files and should not contain anything that directly or indirectly imports semantic files.

**Comment by thewilsonator** [comment]
Closing as much of this has been done to move to `expressionsem.d` and other files by PRs from Razvan and myself and this has a large set of merge conflicts.

**Pull request closed without merging**